### PR TITLE
PERF: Introducing HashTables for datatypes with 8,16 and 32 bits

### PR DESCRIPTION
**Pull request by realead**
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Starts work on #33287

Proof of concept for 32bit/16bit hash tables. If works it should become blue print for UInt32/Int16/UInt16/Float32-HashTables.

### CODE REVIEW

**Comment by realead** [comment]
FYI, @jbrockmendel 

This is  my idea, how adding further HashMap-versions could work. It probably needs more (direct) tests of HashTables, because right now they are tested somewhat indirectly via `test_algos.py` and  a lot of infrastructure is missing for testing the functionality of new versions.

**Comment by jbrockmendel** [comment]
Definitely needs more robust testing, but this looks like a really good start.  Thanks for following upon this.

**Comment by jbrockmendel** [comment]
couple of ideas for either follow-ups or to get testing for "free":

in core.algorithms add Int32HashTable to _hashtables dict and update algorithms._ensure_data to avoid casting int32 to int64 

in _libs.index_class_helper update hashtable_name for int32 (maybe also int16 and int8)

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 1]
```diff
@@ -256,7 +286,9 @@ cdef class HashTable:
 # name, dtype, float_group, default_na_value
 dtypes = [('Float64', 'float64', True, 'np.nan'),
           ('UInt64', 'uint64', False, 0),
-          ('Int64', 'int64', False, 'NPY_NAT')]
+          ('Int64', 'int64', False, 'NPY_NAT'),
+          ('UInt32', 'uint32', False, 0),
+          ('Int32', 'int32', False, 0)]
```
Comment (jreback) [inline pandas/_libs/hashtable_class_helper.pxi.in:291]: this migth be a problem not sure

**Review by jreback — COMMENTED** [review]
wow this looks pretty good. cc @jbrockmendel

File: pandas/_libs/hashtable_class_helper.pxi.in (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/hashtable_class_helper.pxi.in:291]: are you referring to the 0 here?

i looked at this and dont think its an issue, bc it is only a placeholder that never gets used

File: pandas/_libs/src/klib/khash.h [hunk 2]
```diff
@@ -591,6 +591,9 @@ PANDAS_INLINE khint_t __ac_Wang_hash(khint_t key)
 #define KHASH_MAP_INIT_INT(name, khval_t)								\
 	KHASH_INIT(name, khint32_t, khval_t, 1, kh_int_hash_func, kh_int_hash_equal)
 
+#define KHASH_MAP_INIT_UINT(name, khval_t)								\
```
Comment (jbrockmendel) [inline pandas/_libs/src/klib/khash.h:594]: can you add a comment about why the int32 version can be reused verbatim for the uint32 case

File: pandas/tests/libs/test_hashtable.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        table.set_item(index, 21)
+        assert index in table
+        assert index + 1 in table
+        assert len(table) == 2
+        table.get_item(index) == 21
+        table.get_item(index + 1) == 41
+        assert index + 2 not in table
+        with pytest.raises(KeyError) as excinfo:
+            table.get_item(index + 2)
+        assert str(index + 2) in str(excinfo.value)
```
Comment (jbrockmendel) [inline pandas/tests/libs/test_hashtable.py:44]: theres a lot going on in this test.  can it either be split up into smaller pieces, or at least have some whitespace clarify what assertions are connected to what calls

File: pandas/_libs/src/klib/khash.h (hunk 2 above)
Comment (realead) [reply pandas/_libs/src/klib/khash.h:594]: Actually, it is another way around: a comment is needed why int32-version uses verbatim the uint32 case:) `khint32_t` is an unsigned int: https://github.com/pandas-dev/pandas/blob/4cfa97a1808f31c8235d513a7aded32f87f8ae89/pandas/_libs/src/klib/khash.h#L119 so it only naturally to use it for uint32-version

The trick for int32 is that the conversion from signed int to unsigned int is well defined by the standard (see e.g. https://stackoverflow.com/a/50632), thus we can safely pass an int as key (and we don't have something like "get_key", thus it is never casted back to signed int32, which would be implementation defined behavior (and thus not the original value).

Why it is done this way? For example, pandas has tried to improve the situation for int64: https://github.com/pandas-dev/pandas/blob/4cfa97a1808f31c8235d513a7aded32f87f8ae89/pandas/_libs/src/klib/khash.h#L125-L126

The problem now is, that for signed int64 there is implementation defined behavior (>>) and undefined behavior (<<) for negative or "sufficiently large" keys in the hash function: https://github.com/pandas-dev/pandas/blob/4cfa97a1808f31c8235d513a7aded32f87f8ae89/pandas/_libs/src/klib/khash.h#L411

as `key` is now a signed int (see e.g. https://stackoverflow.com/a/4009922). 

I would be surprised to see a (sane) compiler which wouldn't do "the right thing" on x86/x86_64, but still...

The hash-function should be probably be

```
#define kh_int64_hash_func(key) (khint32_t)(((khint64_t)(key))>>33^((khint64_t)(key))^((khint64_t)(key))<<11)
```

just to be safe.

File: pandas/_libs/hashtable_class_helper.pxi.in (hunk 1 above)
Comment (realead) [reply pandas/_libs/hashtable_class_helper.pxi.in:291]: This field  (`default_na_value`) is probably not really needed.

 Just set `na_value2` to 0 here: https://github.com/pandas-dev/pandas/blob/4cfa97a1808f31c8235d513a7aded32f87f8ae89/pandas/_libs/hashtable_class_helper.pxi.in#L410
and drop the whole else-branch here: https://github.com/pandas-dev/pandas/blob/4cfa97a1808f31c8235d513a7aded32f87f8ae89/pandas/_libs/hashtable_class_helper.pxi.in#L432-L433

File: pandas/tests/libs/test_hashtable.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/libs/test_hashtable.py:53]: thanks

File: pandas/_libs/src/klib/khash.h (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/src/klib/khash.h:605]: > Actually, it is another way around: a comment is needed why int32-version uses verbatim the uint32 case:)

fine by me as long as its clear to the next reader

**Comment by jbrockmendel** [comment]
be still my beating heart you got the 8s and the 16s too!

**Comment by realead** [comment]
@jreback @jbrockmendel 

I've added also 8bit versions: mostly for the sake of uniformity - otherwise there is always a special case. While using 8bit hashtable is better than converting to 16bit-datatype and using 16bit hash-table, it is still an overkill. If important, later on a special version could be written for 8bit data types.

Apart from some polishing and small(ish) test-issues, I consider this PR complete. Once it is merged, the upcasting can be dropped at one place after another. I'm however not really a pandas power user and cannot estimate the necessary work and hope somebody will be able to help.

One might not be satisfied with the simple hash-function used (identity), but it is more or less consistent with functions we use for (u)int64. The comments say: https://github.com/pandas-dev/pandas/blob/4cfa97a1808f31c8235d513a7aded32f87f8ae89/pandas/_libs/src/klib/khash.h#L62-L63

So if there are some problems, we could switch to Wang's or murmur2-hash.

File: pandas/tests/libs/test_hashtable.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        keys, counts = value_count(values, False)
+        tm.assert_numpy_array_equal(np.sort(keys), expected)
+        assert np.all(counts == 5)
+
+    def test_duplicated_first(self, dtype, type_suffix):
+        N = 100
+        duplicated = get_ht_function("duplicated", type_suffix)
+        values = np.repeat(np.arange(N).astype(dtype), 5)
+        result = duplicated(values)
+        expected = np.ones_like(values, dtype=np.bool)
```
Comment (jbrockmendel) [inline pandas/tests/libs/test_hashtable.py:187]: looks like these need to be np.bool_ instead of np.bool

**Comment by TomAugspurger** [comment]
@realead can you measure what this does to the size of the distribution? `python setup.py bdist_wheel` should do the trick.

**Comment by realead** [comment]
@TomAugspurger  it adds about 0.7-0.8MB to the wheel.

This PR:
``` 
31.7 Nov 20 22:36 pandas-1.2.0.dev0+1209.g0b9d94a-cp38-cp38-linux_x86_64.whl
7,4M Nov 20 22:24 hashtable.cpython-38-x86_64-linux-gnu.so*
```

master:
```
31M Nov 20 22:51 pandas-1.2.0.dev0+1191.g8d1b8ab-cp38-cp38-linux_x86_64.whl
3.8M Nov 20 22:42 hashtable.cpython-38-x86_64-linux-gnu.so*
```

**Comment by realead** [comment]
@jreback  @jbrockmendel 

I could not help myself and sneaked some changes in, which are not really necessary:

  - for the sake of consistency and to avoid undefined behavior, using unsigned ints also for 64bit integers. This is somewhat hacky (but the way khash has done it in the first place) See also this comment: https://github.com/pandas-dev/pandas/pull/37920#discussion_r527191378. Commit: https://github.com/pandas-dev/pandas/pull/37920/commits/897512d5eba6076340ca4e61068816ab54289f82

 - adding dependency on khash.h, otherwise changing it doesn't trigger rebuild (probably anything cimporting `khash.pxd` (e.g. via `cimport hashtable.pxd`) should have it as well, but I'm not sure). Commit:  https://github.com/pandas-dev/pandas/pull/37920/commits/17a3feed54890d806e25cd731e9559be60d27366

 - getting rid of unnecessary `default_na_value`, see this comment: https://github.com/pandas-dev/pandas/pull/37920#discussion_r526543906  Commit: https://github.com/pandas-dev/pandas/pull/37920/commits/3a4c2bcff2de207b3d993b3d5ec5dd7d2da291df

You can decide whether you want to keep some of them in this PR, or rather in as a new PR or just to drop.

File: pandas/_libs/hashtable_class_helper.pxi.in (hunk 1 above)
Comment (jreback) [reply pandas/_libs/hashtable_class_helper.pxi.in]: yeah that was my comment, but agree not germane for now.

**Comment by jreback** [comment]
> @TomAugspurger it adds about 0.7-0.8MB to the wheel.
> 
> This PR:
> 
> ```
> 31.7 Nov 20 22:36 pandas-1.2.0.dev0+1209.g0b9d94a-cp38-cp38-linux_x86_64.whl
> 7,4M Nov 20 22:24 hashtable.cpython-38-x86_64-linux-gnu.so*
> ```
> 
> master:
> 
> ```
> 31M Nov 20 22:51 pandas-1.2.0.dev0+1191.g8d1b8ab-cp38-cp38-linux_x86_64.whl
> 3.8M Nov 20 22:42 hashtable.cpython-38-x86_64-linux-gnu.so*
> ```

not nothing, but nbd

**Review by jreback — APPROVED** [review]
lgtm. cc @jbrockmendel

**Comment by jbrockmendel** [comment]
LGTM

Travis failure is an unrelated frame plotting test that im seeing locally too

**Comment by jreback** [comment]
thanks @realead very nice!

**Comment by jbrockmendel** [comment]
This is exciting thanks @realead

**Pull request merged**